From edc3015ddcd59ac201b3c243a3ea931f7c05b837 Mon Sep 17 00:00:00 2001
From: Martin Thuemmler <martin.thuemmler@uni-jena.de>
Date: Mon, 10 Nov 2025 10:42:07 +0100
Subject: [PATCH] memory bug fix, when more projection than wannier functions
 are specified

---
 src/wannierise.F90 | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/wannierise.F90 b/src/wannierise.F90
index 2f6b164c..38b754df 100644
--- a/src/wannierise.F90
+++ b/src/wannierise.F90
@@ -413,7 +413,7 @@ contains
 
     ! initialise rguide to projection centres (Cartesians in units of Ang)
     if (wann_control%guiding_centres%enable) then
-      do n = 1, num_proj
+      do n = 1, min(num_wann, num_proj)
         call utility_frac_to_cart(wann_control%guiding_centres%centres(:, n), rguide(:, n), &
                                   real_lattice)
       enddo
-- 
2.36.0

