commit 70a5db9cee9a59d6074ac20d443dbcf00924eb58
Author: Seung-Ju Hong <sjhong6230@snu.ac.kr>
Date:   Mon Apr 29 03:20:38 2024 +0900

    Fixed bug in writing checkpoint file

diff --git a/src/wannier_prog.F90 b/src/wannier_prog.F90
index b6fe4282..f9163c1e 100644
--- a/src/wannier_prog.F90
+++ b/src/wannier_prog.F90
@@ -645,19 +645,15 @@ contains
       return
     endif
 
-    u(:, :, :) = 0.d0
-    uopt(:, :, :) = 0.d0
+    u(:, :, :) = common_data%u_matrix
+    uopt(:, :, :) = common_data%u_opt
     m(:, :, :, :) = 0.d0
 
     do ikl = 1, nkrank
       ikg = global_k(ikl)
-      u(:, :, ikg) = common_data%u_matrix(:, :, ikl)
-      uopt(:, :, ikg) = common_data%u_opt(:, :, ikl)
       m(:, :, :, ikg) = common_data%m_matrix_local(1:nw, 1:nw, :, ikl)
     enddo
 
-    call comms_reduce(u(1, 1, 1), nw*nw*nk, 'SUM', error, common_data%comm)
-    call comms_reduce(uopt(1, 1, 1), nb*nw*nk, 'SUM', error, common_data%comm)
     call comms_reduce(m(1, 1, 1, 1), nw*nw*nn*nk, 'SUM', error, common_data%comm)
     if (allocated(error)) then
       call prterr(error, ierr, istdout, istderr, common_data%comm)
